src/storage/hummock_sdk/src/slice_transform.rs

```diff
@@ -0,0 +1,436 @@
+            .deserialize_prefix_len_with_column_indices(
+                pk,
+                (0..self.read_pattern_prefix_column as usize).collect(),
+            )
+            .unwrap();
+
+        let prefix_len = TABLE_PREFIX_LEN + VIRTUAL_NODE_SIZE + pk_prefix_len;
+        &full_key[0..prefix_len]
+    }
+}
+
+impl SchemaSliceTransform {
+    fn new(table_catalog: &Table) -> Self {
+        assert_ne!(0, table_catalog.read_pattern_prefix_column);
+
+        // column_index in pk
+        let pk_indices: Vec<usize> = table_catalog
+            .order_key
+            .iter()
+            .map(|col_order| col_order.index as usize)
+            .collect();
+
+        let data_types = pk_indices
+            .iter()
+            .map(|column_idx| &table_catalog.columns[*column_idx])
+            .map(|col| ColumnDesc::from(col.column_desc.as_ref().unwrap()).data_type)
+            .collect();
+
+        let order_types: Vec<OrderType> = table_catalog
+            .order_key
+            .iter()
+            .map(|col_order| {
+                OrderType::from_prost(
+                    &risingwave_pb::plan_common::OrderType::from_i32(col_order.order_type).unwrap(),
+                )
+            })
+            .collect();
+
+        Self {
+            read_pattern_prefix_column: table_catalog.read_pattern_prefix_column,
+            deserializer: OrderedRowDeserializer::new(data_types, order_types),
+        }
+    }
+}
+
+#[derive(Default)]
+pub struct MultiSliceTransForm {
+    id_to_slice_transform: HashMap<u32, Arc<dyn SliceTransform>>,
+
+    // cached state
+    last_slice_transform_state: Option<(u32, Arc<dyn SliceTransform>)>,
+}
+
+impl MultiSliceTransForm {
+    fn register(&mut self, table_id: u32, slice_transform: Arc<dyn SliceTransform>) {
+        self.id_to_slice_transform.insert(table_id, slice_transform);
+    }
+
+    fn update_state(&mut self, new_table_id: u32) {
+        match self.id_to_slice_transform.get(&new_table_id) {
+            Some(box_slice_transform) => {
+                self.last_slice_transform_state = Some((new_table_id, box_slice_transform.clone()))
+            }
+
+            None => {
+                unreachable!();
+            }
+        }
+    }
+
+    fn transform<'a>(&mut self, full_key: &'a [u8]) -> &'a [u8] {
+        assert!(full_key.len() > TABLE_PREFIX_LEN + VIRTUAL_NODE_SIZE);
+
+        let table_id = get_table_id(full_key).unwrap();
+        match self.last_slice_transform_state.as_ref() {
+            Some(last_slice_transform_state) => {
+                if table_id != last_slice_transform_state.0 {
+                    self.update_state(table_id);
+                }
+            }
+
+            None => {
+                self.update_state(table_id);
+            }
+        }
+
+        self.last_slice_transform_state
+            .as_ref()
+            .unwrap()
+            .1
+            .transform(full_key)
+    }
+}
+
+#[cfg(test)]
+#[expect(clippy::needless_borrow)]
```

Comment:
We should avoid attributing to the whole module. Remove this or add it to a statement.
```suggestion
```